### PR TITLE
[VPlan][NFC] Add cost for `VPWidenMemoryRecipe`.

### PR DESCRIPTION
**Pull request by ElvisWang123**
In this patch, we add the `computeCost()` function for `VPWidenMemoryRecipe`.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 1]
```diff
@@ -2084,6 +2084,37 @@ void VPPredInstPHIRecipe::print(raw_ostream &O, const Twine &Indent,
 }
 #endif
 
+InstructionCost VPWidenMemoryRecipe::computeCost(ElementCount VF,
+                                                 VPCostContext &Ctx) const {
+  Instruction *I = getInstructionForCost(this);
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2089]: This should probably use `Ingredient`

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 2]
```diff
@@ -2084,6 +2084,37 @@ void VPPredInstPHIRecipe::print(raw_ostream &O, const Twine &Indent,
...
+  Type *Ty = ToVectorTy(getElementType(), VF);
+  const Align Alignment = getLoadStoreAlignment(const_cast<Instruction *>(I));
+  const Value *Ptr = getLoadStorePointerOperand(I);
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2092]: Sink closer to use? Is this required? Using the original IR pointer here won't be accurate I think. If needed, please leave a TODO behind to refactor reliance on underlying IR here and explain why it is currently needed.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 3]
```diff
@@ -2084,6 +2084,37 @@ void VPPredInstPHIRecipe::print(raw_ostream &O, const Twine &Indent,
...
+                                      CostKind, OpInfo, I);
+    }
+    if (Reverse)
+      Cost += Ctx.TTI.getShuffleCost(TargetTransformInfo::SK_Reverse,
+                                     cast<VectorType>(Ty), std::nullopt,
+                                     CostKind, 0);
+
+    return Cost;
+  }
+  return Ctx.TTI.getAddressComputationCost(Ty) +
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2113]: maybe handle the simpler case `(!Consecutive)` first with an early return?

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 4]
```diff
@@ -2529,6 +2529,13 @@ class VPWidenMemoryRecipe : public VPRecipeBase {
     llvm_unreachable("VPWidenMemoryRecipe should not be instantiated.");
   }
 
+  /// Get element Type
+  Type *getElementType() const { return getLoadStoreType(&Ingredient); }
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlan.h:2533]: Simpler to just use `getLoadStoreType` directly in `::computeCost`, like other `getLoadStoreXXX` helpers there?

**Review by fhahn — COMMENTED** [review]
This should be effectively NFC, might be good to tag as such

**Review by fhahn — COMMENTED** [review]
With this change, you should also be able to remove `VPWidenMemoryRecipe` from `getInstructionForCost` in `VPlan.cpp`

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 4 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Thanks, I will directly use the `getLoadStoreType`.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 3 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Sure.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 2 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Only ArmTTI is using the underlying IR to return different gather/scatter cost currently.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 1 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Sure.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 5]
```diff
@@ -2079,6 +2079,44 @@ void VPPredInstPHIRecipe::print(raw_ostream &O, const Twine &Indent,
...
+        Ctx.TTI.getOperandInfo(Ingredient.getOperand(0));
+    Cost += Ctx.TTI.getMemoryOpCost(Ingredient.getOpcode(), Ty, Alignment, AS,
+                                    CostKind, OpInfo, &Ingredient);
+  }
+  if (Reverse)
+    Cost +=
+        Ctx.TTI.getShuffleCost(TargetTransformInfo::SK_Reverse,
+                               cast<VectorType>(Ty), std::nullopt, CostKind, 0);
+
+  return Cost;
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2117]: nit: handling the non-reverse case with an early exit may be slightly more compact

```suggestion
  if (!Reverse)
  return Cost;
   
   return 
    Cost +
        Ctx.TTI.getShuffleCost(TargetTransformInfo::SK_Reverse,
                               cast<VectorType>(Ty), std::nullopt, CostKind, 0);
```

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 6]
```diff
@@ -2079,6 +2079,44 @@ void VPPredInstPHIRecipe::print(raw_ostream &O, const Twine &Indent,
...
+      getLoadStoreAlignment(const_cast<Instruction *>(&Ingredient));
+  unsigned AS =
+      getLoadStoreAddressSpace(const_cast<Instruction *>(&Ingredient));
+  TTI::TargetCostKind CostKind = TTI::TCK_RecipThroughput;
+
+  if (!Consecutive) {
+    // TODO: Using the original IR may not be accurate.
+    // Currently, ARM will use the underlying IR to calculate gather/scatter
+    // instruction cost.
+    const Value *Ptr = getLoadStorePointerOperand(&Ingredient);
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2095]: nit: can we assert !Reverse here?

**Review by fhahn — COMMENTED** [review]
With the change it should also be possible to  remove `VPWidenMemoryRecipe` from `getInstructionForCost` in `VPlan.cpp`

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 5 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Updated, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 6 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2095]: Sure.

**Comment by fhahn** [comment]
> With the change it should also be possible to remove `VPWidenMemoryRecipe` from `getInstructionForCost` in `VPlan.cpp`

Still missing?

**Comment by ElvisWang123** [comment]
When removing `VPWidenMemoryRecipe` from the `getInstructionForCost()` will cause test fails.
Because in the `VPRecipeBase::cost()` , there is a check for override the instruction cost by option needs the return from `getInstructionForCost()` .
```
if (UI && ForceTargetInstructionCost.getNumOccurrences() > 0 &&
      RecipeCost.isValid())
    RecipeCost = InstructionCost(ForceTargetInstructionCost);
```

Should we just remove the check of `UI` ?

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 7]
```diff
@@ -2079,6 +2079,46 @@ void VPPredInstPHIRecipe::print(raw_ostream &O, const Twine &Indent,
...
+        Ctx.TTI.getOperandInfo(Ingredient.getOperand(0));
+    Cost += Ctx.TTI.getMemoryOpCost(Ingredient.getOpcode(), Ty, Alignment, AS,
+                                    CostKind, OpInfo, &Ingredient);
+  }
+  if (!Reverse)
+    return Cost;
+
+  return Cost += Ctx.TTI.getShuffleCost(TargetTransformInfo::SK_Reverse,
+                                        cast<VectorType>(Ty), std::nullopt,
+                                        CostKind, 0);
```
Comment (arcbbb) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:2119]: not sure if this is still NFC but we can skip the shuffle cost if it is a store and the store value is uniform

**Comment by fhahn** [comment]
> > > With the change it should also be possible to remove `VPWidenMemoryRecipe` from `getInstructionForCost` in `VPlan.cpp`
> > 
> > 
> > Still missing?
> 
> Remove `VPWidenMemoryRecipe` from `getInstructionForCost` will cause a test fail due to different VF between legacy model. Maybe leave a TODO behind to revisit it when we don't need to compare to the legacy model?

yes that would be great, together with an explanation that VPWidenMemoryRecipe case in getInstructionForCost still being needed for force-target-instruction handling.

**Comment by ElvisWang123** [comment]
> > > > With the change it should also be possible to remove `VPWidenMemoryRecipe` from `getInstructionForCost` in `VPlan.cpp`
> > > 
> > > 
> > > Still missing?
> > 
> > 
> > Remove `VPWidenMemoryRecipe` from `getInstructionForCost` will cause a test fail due to different VF between legacy model. Maybe leave a TODO behind to revisit it when we don't need to compare to the legacy model?
> 
> yes that would be great, together with an explanation that VPWidenMemoryRecipe case in getInstructionForCost still being needed for force-target-instruction handling.

Done, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 8]
```diff
@@ -271,6 +271,12 @@ static Instruction *getInstructionForCost(const VPRecipeBase *R) {
     return dyn_cast_or_null<Instruction>(S->getUnderlyingValue());
   if (auto *IG = dyn_cast<VPInterleaveRecipe>(R))
     return IG->getInsertPos();
+  // Currently the legacy cost model only calculates the instruction cost with
+  // underlying instruction. Removing the WidenMem here will prevent
+  // force-target-instruction-cost overwriting the cost of recipe with
+  // underlying instruction which is inconsistent with the legacy model.
+  // TODO: Remove WidenMem from this function when we don't need to compare to
+  // the latecy model.
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:279]:
```suggestion
  // the legacy model.
```
?

**Review by fhahn — APPROVED** [review]
LGTM,thanks!

**Pull request merged**